### ISSUE TITLE
Pose Data from ARKit

### ISSUE THREAD

**Issue by ianmkim** [magicleap/Atlas#34]
I'm trying to see if I could use the pose data from ARKit with Atlas as suggested by @zmurez-ml in issue #2. However, Atlas is returning either 0 vertex models or completely unrelated irrelevant shapes like the one below (should be a kitchen instead)

![Capture](https://user-images.githubusercontent.com/42292798/103139697-de21ac00-46ac-11eb-8472-178e236db6dd.PNG)

The pose data is generated from ARKit, and the coordinate system is converted from the right-handed convention to the left-handed by multiplying the pose data with this matrix [[1,0,0,0],[0,1,0,0],[0,0,-1,0],[0,0,0,1]] and converted from millimeters to meters.

The images are captured at the rate of 10 frames per second. 

I know this isn't enough information, but I'm not sure what more information would help solve the issue so please do let me know.

Thanks

**Comment by coledea** [comment]
I've got the same problem. Any update on this?

**Comment by HaFred** [comment]
Same here. Maybe we could invite @zmurez-ml, maybe just elaborate on how specifically you did for obtaining pose data. Referring to this https://github.com/magicleap/Atlas/issues/2#issuecomment-677217882.

**Comment by HaFred** [comment]
> I've got the same problem. Any update on this?

I tried with the ARKit data and it comes out as a flat result just like this. Have you solved this? Thanks.

**Comment by coledea** [comment]
Yes, I think I found out, how the pose data from ARKit has to be converted.  This is the transform I apply to the camera transformation matrix returned by the ARKit:
```flip_yz_matrix = np.array([[1, 0, 0], [0, -1, 0], [0, 0, -1]])
 rotation = rotation_matrix_from_euler(camera_transform_euler_angles)
 rotation = rotation.dot(flip_yz_matrix)
 rotation = rotate_around_x(np.pi / 2) @ rotation
 translation = camera_transform_translation
 translation = rotate_around_x(np.pi / 2) @ translation
 final_matrix = np.zeros([3, 4])
 final_matrix[:3, :3] = rotation
 final_matrix[:3, 3] = translation
 final_matrix = np.vstack((final_matrix, [0, 0, 0, 1]))
 ```
 
The camera_transform_euler_angles and the camera_transform_translation come from the ARKit. The rotate_around_x function creates a 3x3 rotation matrix.
Additionally, I had to add 5 to xyz of the translation vector, otherwise, only a corner of the room is reconstructed. This value was just a random guess. I have not yet looked into this (why this shift is necessary and whether 5 works in general or only for this one room I tried to reconstruct. Maybe the positions have to be shifted so that they have a positive value)

**Comment by HaFred** [comment]
Thanks for the reply. Actually, I tried the same conversion as you did before, as in the `kp_reproject.py`. But the reconstruction result with `inference.py` is not so good.
![image](https://user-images.githubusercontent.com/13991298/124921072-9c45cb00-e02a-11eb-8746-d413ae296565.png)
I have some thoughts as follows:

- I compared the quaternions-based and transform-matrix-based methods by ARKit, the pose matrices for these two cases are almost the same, thus I guess the ARKit output [transform](https://developer.apple.com/documentation/arkit/arcamera/2866108-transform) is reliable, with the unit of meter and Y-up.
- Then the conversion code we both used above, used the `rotate_around_x(pi/2)` to convert it in Z-up for Atlas's usage.
- I believe with such conversion, the `final_matrix` should already be the extrinsic matrix, isn't it? Then I am not sure whether the inversion in this line https://github.com/magicleap/Atlas/blob/bee3ddbdb0935cf9aa9fc810e829115c6ab74dc5/atlas/transforms.py#L60 is still needed?
- The snapshot here is the less translation one. I figured comparing to their [sample scene](https://drive.google.com/file/d/13-D7QNVZjj864E768zJ7IWDDqY6agUES/view?usp=sharing) in readme, maybe I need a huge translation data for inference. However, the result is even worse, the `tsdf_pred` only dump 10 edges and vertices (ply with only 1KB).
- I noticed that you mentioned about these
> Additionally, I had to add 5 to xyz of the translation vector, otherwise, only a corner of the room is reconstructed. This value was just a random guess. I have not yet looked into this (why this shift is necessary and whether 5 works in general or only for this one room I tried to reconstruct. Maybe the positions have to be shifted so that they have a positive value)

Is it the `nviews`? I don't get how `middle_pose` works here, maybe you could elaborate on this a little bit?

Thank you very much!

**Comment by HaFred** [comment]
Also, I think 24GB VRAM is sufficient for `voxel_size 224 224 64` reconstruction. Referring to the sample data reconstruction, our custom data here is even smaller, but the inference result cannot achieve the quality of sample data.
I think it may due to the incompatibility of for the pre-trained model the author provided. Since the sample inference data looks close to scannet data (pretrained model), while our custom data might not. So it leads to bad inference result.

**Comment by coledea** [comment]
* The inversion in transforms.py line 60 is necessary, I think. `final_matrix` should be the camera's pose matrix (so it describes position and rotation of the camera in world space), but we need the extrinsics matrix, (aka view matrix) which is the inverse of the camera pose matrix.
* I don't know what `nviews` or `middle_pose` is. I did not find it in the code. When I calculate the bounding volume for my scene (in meters), I get a minimum that is around (-5,-5,-5). After adding 5 to the translation vector of the camera's pose matrix, the whole room is reconstructed correctly (low-quality though). So maybe try to calculate the bounding volume and move the minimum to the origin by adding the absolute value to the translation vector of the camera's pose matrices.
You can calculate the bounding volume for example like this (copied from [here](https://github.com/andyzeng/tsdf-fusion-python)):
```
def rigid_transform(xyz, transform):
    xyz_h = np.hstack([xyz, np.ones((len(xyz), 1), dtype=np.float32)])
    xyz_t_h = np.dot(transform, xyz_h.T).T
    return xyz_t_h[:, :3]

def get_view_frustum(depth_im, cam_intr, cam_pose):
      im_h = depth_im.shape[0]
      im_w = depth_im.shape[1]
      max_depth = np.max(depth_im)
      view_frust_pts = np.array([
        (np.array([0,0,0,im_w,im_w])-cam_intr[0,2])*np.array([0,max_depth,max_depth,max_depth,max_depth])/cam_intr[0,0],
        (np.array([0,0,im_h,0,im_h])-cam_intr[1,2])*np.array([0,max_depth,max_depth,max_depth,max_depth])/cam_intr[1,1],
        np.array([0,max_depth,max_depth,max_depth,max_depth])
      ])
      view_frust_pts = rigid_transform(view_frust_pts.T, cam_pose).T
      return view_frust_pts

  
vol_bnds = np.zeros((3,2))
for i in range(n_imgs):
    # Read depth image and camera pose + intrinsics

    # Compute camera view frustum and extend convex hull
    view_frust_pts = get_view_frustum(depth_image, camera_intrinsics, camera_pose)
    vol_bnds[:,0] = np.minimum(vol_bnds[:,0], np.amin(view_frust_pts, axis=1))
    vol_bnds[:,1] = np.maximum(vol_bnds[:,1], np.amax(view_frust_pts, axis=1))
```

**Comment by HaFred** [comment]
Okay. I have one more question. Noted that in the snippet here the RGBD depth info is used to calculate the bounding, so did you also include depth information for the bounding you did for better inference using atlas? Thanks.

**Comment by coledea** [comment]
If I'm not mistaken, you need the depth to calculate the bounding volume. If you have only camera pose and intrinsics you know nothing about the size of the volume w.r.t the z-axis (assuming the z-axis is standing orthogonal on the camera plane). The maximum depth in one depth image gives you the information, where on the z-axis the bounding volume ends.

**Comment by HaFred** [comment]
Great, I got your point thanks. However, if the calibration for translation vector is needed (adding bounding volume which is calculated by depth image), how come the [sample data](https://drive.google.com/file/d/13-D7QNVZjj864E768zJ7IWDDqY6agUES/view?usp=sharing) can be reconstructed by Atlas without such calibration? I know I can retrieve bounding volume by Kinect in such case, but to use Atlas I wish to eliminate the usage of RGBD camera.

**Comment by coledea** [comment]
In the ScanNet Paper ([Link to Paper](https://arxiv.org/pdf/1702.04405.pdf)) they state: 

> Finally, we use a PCA on the mesh vertices to determine the rotation around the z-axis and translate the scan such that its bounds are within the positive octant of the coordinate system.

So they shift the geometry into the positive octant. This is probably the reason, why we have to shift the ARKit Poses into the positive octant (by adding the minimum of the bounding volume).

**Comment by HaFred** [comment]
> In the ScanNet Paper ([Link to Paper](https://arxiv.org/pdf/1702.04405.pdf)) they state:
> 
> > Finally, we use a PCA on the mesh vertices to determine the rotation around the z-axis and translate the scan such that its bounds are within the positive octant of the coordinate system.
> 
> So they shift the geometry into the positive octant. This is probably the reason, why we have to shift the ARKit Poses into the positive octant (by adding the minimum of the bounding volume).

Hey, sorry for the late reply. I would like to know how did you measure **the depth** together with camera pose & intrinsics for your scene particularly? Are you using ARKit on iPhone12 or Kinect or some other depth camera?

**Comment by coledea** [comment]
I used the ARKit on an IPad Pro (3rd Generation), which has an lidar sensor.
